src/backend/sev/builder.rs

```diff
@@ -10,33 +10,75 @@ use crate::backend::kvm::mem::Region;
 
 use std::convert::TryFrom;
 use std::sync::{Arc, RwLock};
+use std::{thread, time};
 
-use anyhow::{Context, Error, Result};
+use anyhow::{Context, Error};
 use kvm_ioctls::Kvm;
 use mmarinus::{perms, Map};
 use primordial::Page;
+use rand::{thread_rng, Rng};
 use sallyport::elf::pf::snp::{CPUID, SECRETS};
 use x86_64::VirtAddr;
 
+const SEV_RETRIES: usize = 3;
+const SEV_RETRY_SLEEP_MS: u64 = 500;
+
 pub struct Builder {
     kvm_fd: Kvm,
     launcher: Launcher<Started, Firmware>,
     regions: Vec<Region>,
     sallyports: Vec<Option<VirtAddr>>,
 }
 
+fn retry<O, F>(func: F) -> anyhow::Result<O>
+where
+    F: Fn() -> anyhow::Result<O>,
+{
+    let mut retries = SEV_RETRIES;
+    let mut rng = thread_rng();
+    loop {
+        match func() {
+            Err(e) if retries > 0 => {
+                retries -= 1;
+                eprintln!(
+                    "Error {:#?}.\nRetry {} of {}.",
+                    e,
+                    SEV_RETRIES - retries,
+                    SEV_RETRIES
+                );
+                let millis =
+                    time::Duration::from_millis(SEV_RETRY_SLEEP_MS + rng.gen::<u8>() as u64);
+                thread::sleep(millis);
+                continue;
+            }
+            Err(e) => {
+                return Err(e);
+            }
+            Ok(o) => {
+                return Ok(o);
+            }
+        }
+    }
+}
+
 impl TryFrom<super::super::kvm::config::Config> for Builder {
     type Error = Error;
 
-    fn try_from(_config: super::super::kvm::config::Config) -> Result<Self> {
-        let kvm_fd = Kvm::new().context("Failed to open '/dev/kvm'")?;
-        let vm_fd = kvm_fd
-            .create_vm()
-            .context("Failed to create a virtual machine")?;
+    fn try_from(_config: super::super::kvm::config::Config) -> anyhow::Result<Self> {
+        let (kvm_fd, launcher) = retry(|| {
+            // try to open /dev/sev and start the Launcher several times
+
+            let kvm_fd = Kvm::new().context("Failed to open '/dev/kvm'")?;
+            let vm_fd = kvm_fd
+                .create_vm()
+                .context("Failed to create a virtual machine")?;
+
+            let sev = retry(|| Firmware::open().context("Failed to open '/dev/sev'"))?;
 
-        let sev = Firmware::open().context("Failed to open '/dev/sev'")?;
+            let launcher = Launcher::new(vm_fd, sev).context("SNP Launcher init failed")?;
 
-        let launcher = Launcher::new(vm_fd, sev).context("SNP Launcher init failed")?;
+            Ok((kvm_fd, launcher))
```

Comment:
not that it really matters, but I actually preferred the original version with a map
Not entirely sure what we gain with this refactor commit

Reply:
I find it poor for readability to smuggle the returned tuple inside a map, especially when what should logically be the "returned" thing is immediately followed by a error handling for an operation unrelated to the return. I'd prefer to keep the launcher creation logic self-contained, and return the final tuple as its own statement.

The reason that the map was used in the first place was for type-system-tetris to avoid a type annotation which arose from being overly generic on the return type of `retry`. This was itself an instance of poor style, since we had both `std::result::Result` and `anyhow::Result` in scope as simply `Result` and were relying on the number of type parameters to distinguish them, but that makes for poor readability. In any case, we didn't need to be using `std::result::Result` as the return type of `retry` since we only use `retry` with `anyhow::Result`. In general it's just good style not to have two types in scope with the exact same name; given usual Rust naming conventions this usually only occurs for Result types (and their variants), so it is encouraged to use them namespaced as `foo::Result` to reduce confusion. And by fixing this style nit (by just using `anyhow::Result` everywhere) we also remove the impetus for using `.map` to avoid the type inference failure.